eth-rpc-adapter/src/server.ts

```diff
@@ -1,77 +1,227 @@
-import { EvmRpcProvider } from '@acala-network/eth-providers';
-import HTTPServerTransport from './transports/http';
-import WebSocketServerTransport from './transports/websocket';
-import { Eip1193Bridge } from './eip1193-bridge';
-import { RpcForward } from './rpc-forward';
+import http, { ServerOptions } from 'http';
+import WebSocket from 'ws';
+import cors from 'cors';
+import { json as jsonParser } from 'body-parser';
+import connect, { HandleFunction } from 'connect';
+import { logger } from './logger';
+import { errorHandler } from './middlewares';
+import { InvalidRequest, MethodNotFound, BatchSizeError } from './errors';
 import { Router } from './router';
-import { version } from './_version';
-import { parseOptions } from './utils';
+import { DataDogUtil } from './utils';
 
-export async function start(): Promise<void> {
-  console.log('starting server ...');
+export interface EthRpcServerOptions extends ServerOptions {
+  port: number;
+  batchSize: number;
+  middleware?: HandleFunction[];
+  cors?: cors.CorsOptions;
+}
+
+export interface JsonRpcRequest {
+  jsonrpc: string;
+  id: string;
+  method: string;
+  params: any[] | Record<string, unknown>;
+}
+
+export interface JsonRpcError {
+  code: number;
+  message: string;
+  data?: any;
+}
+
+export interface JsonRpcResponse {
+  jsonrpc: string;
+  id: string | null;
+  result?: any;
+  error?: JsonRpcError;
+}
+
+export default class EthRpcServer {
+  private static defaultCorsOptions = { origin: '*' };
+  private server: http.Server;
+  private wss: WebSocket.Server;
+  private options: EthRpcServerOptions;
+  public routers: Router[] = [];
+
+  constructor(options: EthRpcServerOptions) {
+    this.options = options;
+
+    const app = connect();
 
-  const opts = parseOptions();
-  const provider = EvmRpcProvider.from(opts.endpoints.split(','), {
-    safeMode: opts.safeMode,
-    localMode: opts.localMode,
-    richMode: opts.richMode,
-    verbose: opts.verbose,
-    subqlUrl: opts.subqlUrl,
-    maxBlockCacheSize: opts.maxBlockCacheSize,
-    storageCacheSize: opts.storageCacheSize
-  });
+    options.middleware?.forEach((mw) => app.use(mw));
+    const corsOptions = options.cors ?? EthRpcServer.defaultCorsOptions;
+    app.use(cors(corsOptions));
+    app.use(jsonParser({ limit: '1mb' }));
+    app.use(this.httpHandler.bind(this));
+    app.use(errorHandler);
 
-  const bridge = new Eip1193Bridge(provider);
+    this.server = http.createServer(app);
 
-  const rpcForward = opts.forwardMode ? new RpcForward(provider) : undefined;
+    this.wss = new WebSocket.Server({ server: this.server });
```

Comment:
Confirmed with bryan that it would be nice to have a switch to control whether to start the websocket server or not.

Reply:
sounds good, I can add it in a separate commit to prevent this PR to get over complicated. By the way what's the use case for only allowing http but not ws port?

Reply:
I think it's based on performance considerations, and it doesn't really matter.

Reply:
cool, I can add it along with #555